src/textual/widgets/_data_table.py

```diff
@@ -1767,58 +1845,113 @@ def _render_cell(
+            component_style, post_style = self._get_styles_to_render_cell(
+                is_header_cell,
+                is_row_label_cell,
+                is_fixed_style_cell,
+                hover,
+                cursor,
+                self.show_cursor,
+                self._show_hover_cursor,
+                self.cursor_foreground_priority == "css",
+                self.cursor_background_priority == "css",
             )
 
+            if is_header_cell:
+                options = self.app.console.options.update_dimensions(
+                    width, self.header_height
+                )
+            else:
+                row = self.rows[row_key]
+                # If an auto-height row hasn't had its height calculated, we don't fix
+                # the value for `height` so that we can measure the height of the cell.
+                if row.auto_height and row.height == 0:
+                    options = self.app.console.options.update_width(width)
+                else:
+                    options = self.app.console.options.update_dimensions(
+                        width, row.height
+                    )
             lines = self.app.console.render_lines(
                 Styled(
                     Padding(cell, (0, 1)),
                     pre_style=base_style + component_style,
-                    post_style=post_foreground + post_background,
+                    post_style=post_style,
                 ),
-                self.app.console.options.update_dimensions(width, height),
+                options,
             )
 
             self._cell_render_cache[cell_cache_key] = lines
 
         return self._cell_render_cache[cell_cache_key]
 
+    @functools.lru_cache(
+        maxsize=512
+    )  # 9 Boolean arguments = 512 possible combinations.
+    def _get_styles_to_render_cell(
+        self,
+        is_header_cell: bool,
+        is_row_label_cell: bool,
+        is_fixed_style_cell: bool,
+        hover: bool,
+        cursor: bool,
+        show_cursor: bool,
+        show_hover_cursor: bool,
+        has_css_foreground_priority: bool,
+        has_css_background_priority: bool,
+    ) -> tuple[Style, Style]:
+        """Auxiliary method to compute styles used to render a given cell.
+
+        Args:
+            is_header_cell: Is this a cell from a header?
+            is_row_label_cell: Is this the label of any given row?
+            is_fixed_style_cell: Should this cell be styled like a fixed cell?
+            hover: Does this cell have the hover pseudo class?
+            cursor: Is this cell covered by the cursor?
+            show_cursor: Do we want to show the cursor in the data table?
+            show_hover_cursor: Do we want to show the mouse hover when using the keyboard
+                to move the cursor?
+            has_css_foreground_priority: `self.cursor_foreground_priority == "css"`?
+            has_css_background_priority: `self.cursor_background_priority == "css"`?
```

Comment:
Yesterday we talked about rhetorical questions as docstrings, but given that these are just a bunch of Boolean values, this actually looks reasonable to me.